### DISCUSSION TITLE
Truncation suggestion

### DISCUSSION THREAD

**Discussion by ishah-twilio** [twilio-labs/paste#1347] [Q&A] [answered]
Paste guidelines suggest that -- "Truncation should NOT be used on page headers, titles, labels, error messages, validation messages, notifications, or SIDs."

Can team have any guidelines on for a report that has 3 different types of SID as data entry? Our existing reports truncates SID in the center and we like to continue using that style. Also SID is not important to users. Users wants and needs ICCID for reference and that is not truncating.

<img width="830" alt="Screen Shot 2021-04-07 at 1 19 05 PM" src="https://user-images.githubusercontent.com/69016180/113929089-777c5600-97a4-11eb-8322-befa2ffaab20.png">

**Comment by serifluous** [comment, marked as answer]
@ishah-twilio Our truncation guideline around SIDs is mostly because there are a number of use cases where customers need to copy-and-paste SIDs into other parts of the Twilio product, command lines, etc., as well as verify that they're identifying the right object with a SID.

However if you know that one of the SIDs you have listed isn't important to users, are you able to remove it? If the Command SID is the only unique identifier for that table row, it's also probably fine to truncate if you know that customers will never need to see or interact with the whole string of text.

**Reply by ghostling** [reply]
@serifluous Can you provide some guidance on:

> Our existing reports truncates SID in the center and we like to continue using that style.

In my experience with how humans at Twilio interact with SIDs is that scan for the first few characters, some scan for the last few characters. Can we consider that for a use case of the Truncate component?

**Reply by serifluous** [reply]
@ghostling sure! We can definitely update [the guidelines](https://paste.twilio.design/components/truncate). How's something like this?

> Good use cases for truncation include:
> - Breadcrumbs
> - Long URL links
> - SIDs used only for identification or reference purposes (for example, if the customer needs only to see the last or first few unique characters)
>
> Truncation should NOT be used on page headers, titles, labels, error messages, validation messages, notifications, or SIDs that customers need to copy and paste.

**Reply by ghostling** [reply]
Yes that's fine, but currently the [Truncate](https://paste.twilio.design/components/truncate/#paste-docs-content-area) component only cuts off the end and appends ellipsis. I don't see a way to truncate the **middle** of a long string. 🙁

**Reply by serifluous** [reply]
Ah, got it. Let me chat with the team!

For some context, is there a reason customers need to see both the first and last few characters, and not just the first or last character?

**Reply by ghostling** [reply]
I just have anecdotal evidence from the past 5 years of watching Twilions use SIDs internally and have been watching people use either first few or last few. Personally I use last few 🤷‍♀️ , but would be something good to key an eye out for...

**Reply by vankhoawin** [reply]
>  is there a reason customers need to see both the first and last few characters, and not just the first or last character?

for me (and i'm sure for others consciously or subconsciously), it's to ensure uniqueness that this is the SID you're looking for. Github agrees that truncating is fine, if there's enough characters to guarantee that uniqueness :)

https://github.com/twilio-labs/paste/commit/f10e615097a32bc8434523e6c269773bb96bad25 (full URL)
https://github.com/twilio-labs/paste/commit/f10e615097a32bc84345 (half truncated URL)
https://github.com/twilio-labs/paste/commit/f10e61 (6 characters)
https://github.com/twilio-labs/paste/commit/f10e (this one 404s)

**Comment by ishah-twilio** [comment]
For IoT Super SIM usecases, Customers never really use the SIM SID at all. Not to look at for any reason we discovered in the phase of research. We have interviewed 20+ customers in the year of 2020 for various usecases and all of them have said give me ICCID and don't force me to use the SID. SIM SID it Twilio's internal identifier and ICCID is the most commonly used identifier for SIM in the industry. 

It is more likely that customer will copy and paste the ICCID than the SIM SID. For this table -- Command logs feature Command SID is the unique identifier. So it should/could be displayed in full. Though during the usability studies customers were ok to scan through the truncated list and click on the link for details. 

We like to truncate the SIM SID in the middle to allow customer to scan the uniqueness if necessary. And we will monitor the support tickets and customer ask in future research studies if necessary. 

User will more likely copy the ICCID than SIM SID. To avoid the accidental clicking on the link we like to not link the ICCID. 

Details page displays the full SIM ID / ICCID.

**Comment by serifluous** [comment]
@ishah-twilio @ghostling @vnguyen94 Thanks for all the input! Since the current component is truncating based on width, rather than character count, it's going to take us a little more digging into use cases to figure out:

1. The number of characters the component shows
2. How customizable we make that number (if at all)

^ This is true for both truncation at the end and start of a string, but gets even trickier with middle truncation. For now, we recommend applying your own truncation for your product, and we'll take note of this use case for our next iteration of the Truncation component.
